### PR TITLE
Add multi-object tracking dabble node

### PR DESCRIPTION
**Pull request by jovi-s**
Add IOU_Tracker and MOSSE (from OpenCV) trackers that can be used with peekingduck's object detectors thus being multi-class.

Closes #476

### CODE REVIEW

File: docs/source/use_cases/multi_object_tracking.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
+# Multi Object Tracking
+
+## Overview
+
+Multi Object Tracking (MOT) is the task of taking an initial set of object detections, creating a unique ID for each of the initial detections, and then tracking each of the objects as they move around frames in a video, maintaining the ID assignment.
```
Comment (diwei-tan) [inline docs/source/use_cases/multi_object_tracking.md:5]:
```suggestion
Multi Object Tracking (MOT) is the task of detecting unique objects and tracking them as they move across frames in a video.
```

File: peekingduck/pipeline/nodes/dabble/utils/load_tracker.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+from .tracking_files.iou_tracking import IOUTracking
+from .tracking_files.opencv_tracking import OpenCVTracker
+
+
+class LoadTracker:  # pylint: disable=too-few-public-methods
+    """Loads chosen tracker node."""
+    def __init__(self, tracking_type: str) -> None:
+        super().__init__()
+        self.logger = logging.getLogger(__name__)
+        if tracking_type == "mosse":
```
Comment (diwei-tan) [inline peekingduck/pipeline/nodes/dabble/utils/load_tracker.py:30]: Its cleaner to put in a dictionary say {'iou': IOUTracker, 'mouse': MOSSETracker} and pull out. This way if eventually the number of tracker models increase we can just add to the registry.

File: peekingduck/pipeline/nodes/dabble/utils/tracking_files/iou_tracker/iou_tracker.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+"""
+Core IOU Tracker
```
Comment (diwei-tan) [inline peekingduck/pipeline/nodes/dabble/utils/tracking_files/iou_tracker/iou_tracker.py:36]:
```suggestion
IOU Tracker
```

File: peekingduck/pipeline/nodes/dabble/utils/tracking_files/iou_tracker/track.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+"""
+Track class
+"""
+
+from typing import Any, Tuple, Union
+import numpy as np
+
+class Track:  # pylint: disable=too-many-instance-attributes
+    """
+    Track containing attributes to track various objects.
```
Comment (diwei-tan) [inline peekingduck/pipeline/nodes/dabble/utils/tracking_files/iou_tracker/track.py:44]:
```suggestion
    Class containing attributes to track various objects.
```

File: peekingduck/pipeline/nodes/dabble/utils/tracking_files/iou_tracker/track.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+            self.bbox[3], self.detection_confidence, -1, -1, -1
+        )
+        return mot_tuple
+
+    # pylint: disable=no-self-use
+    def predict(self) -> None:
+        """Implement to prediction the next estimate of track."""
+        # pylint: disable=notimplemented-raised
+        # pylint: disable=raising-bad-type
+        raise NotImplemented
```
Comment (diwei-tan) [inline peekingduck/pipeline/nodes/dabble/utils/tracking_files/iou_tracker/track.py:171]: Why not just remove this if it is not used and not implemented?

File: peekingduck/pipeline/nodes/dabble/utils/tracking_files/iou_tracker/tracker.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+            outputs = self._get_tracks(self.tracks)
+            return outputs
+
+        detections = Tracker.preprocess_input(bboxes, class_ids, detection_scores)
+
+        track_ids = list(self.tracks.keys())
+
+        updated_tracks, updated_detections = [], []
+
+        # pylint: disable=len-as-condition
```
Comment (diwei-tan) [inline peekingduck/pipeline/nodes/dabble/utils/tracking_files/iou_tracker/tracker.py:219]: Please discuss during standup whether we should disable or rewrite

File: peekingduck/pipeline/nodes/dabble/utils/tracking_files/iou_tracker/tracker.py [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+        """
+
+        new_bboxes = np.array(bboxes, dtype='int')
+        new_class_ids = np.array(class_ids, dtype='int')
+        new_detection_scores = np.array(detection_scores)
+
+        new_detections = list(zip(new_bboxes, new_class_ids, new_detection_scores))
+        return new_detections
+
+    # pylint: disable=too-many-locals
```
Comment (diwei-tan) [inline peekingduck/pipeline/nodes/dabble/utils/tracking_files/iou_tracker/tracker.py:175]: Please discuss during standup whether we should disable or rewrite

File: tests/use_cases/systest_object_tracking.yml [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+- input.recorded:   
+    input_dir: "images/testing"
+- model.yolo:
+    model_type: "v4tiny"
+- dabble.fps
+- dabble.tracking:
+    tracking_type: "iou"
+- draw.tag
+- draw.bbox
+- draw.legend
```
Comment (diwei-tan) [inline tests/use_cases/systest_object_tracking.yml:11]:
```suggestion
- draw.legend

```

File: requirements.txt [hunk 9]
```diff
@@ -1,5 +1,5 @@
 click == 7.1.2
-opencv-python == 4.1.2.30
+opencv-contrib-python == 4.1.2.30
```
Comment (diwei-tan) [inline requirements.txt:2]: Why is this change made? What is the difference between contrib and the normal one?

File: peekingduck/pipeline/nodes/dabble/utils/tracking_files/opencv_tracking.py [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+            tracker = cv2.TrackerMedianFlow_create()
+        elif tracker_type == tracker_types[5]:
+            tracker = cv2.TrackerGOTURN_create()
+        elif tracker_type == tracker_types[6]:
+            tracker = cv2.TrackerMOSSE_create()
+        elif tracker_type == tracker_types[7]:
+            tracker = cv2.TrackerCSRT_create()
+        else:
+            tracker = None
+        return tracker
```
Comment (diwei-tan) [inline peekingduck/pipeline/nodes/dabble/utils/tracking_files/opencv_tracking.py:157]: Is this needed? There seems to be overlapping files? Can you check? Thanks. Please remove redundant files (only leave mosse and iou for now)

**Review by diwei-tan — COMMENTED** [review]
Please reply on comments. Let's discuss.

File: peekingduck/pipeline/nodes/dabble/utils/load_tracker.py (hunk 2 above)
Comment (jovi-s) [reply peekingduck/pipeline/nodes/dabble/utils/load_tracker.py:30]: This is a good idea, I will implement this

File: docs/source/use_cases/multi_object_tracking.md (hunk 1 above)
Comment (jovi-s) [reply docs/source/use_cases/multi_object_tracking.md:5]: Will implement this suggestion

File: docs/source/use_cases/multi_object_tracking.md [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+<img src="https://raw.githubusercontent.com/aimakerspace/PeekingDuck/dev/images/readme/yolo_demo.gif" width="70%">
+
+#### 2. MOSSE (OpenCV)
+
+Minimum Output Sum of Squared Error (MOSSE) uses an adaptive correlation for object tracking which produces stable correlation filters when initialized using a single frame. MOSSE tracker is robust to variations in lighting, scale, pose, and non-rigid deformations. It also detects occlusion based upon the peak-to-sidelobe ratio, which enables the tracker to pause and resume where it left off when the object reappears. MOSSE tracker also operates at a higher fps. To add to the positives, it is also very easy to implement, is as accurate as other complex trackers and much faster. But, on a performance scale, it lags behind the deep learning based trackers.
```
[the remark this answers is not in the document]
Comment (jovi-s) [reply docs/source/use_cases/multi_object_tracking.md:33]: Will implement this suggestion

File: docs/source/use_cases/multi_object_tracking.md [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+The bounding boxes detected in the first frame are used to initialize a single tracker instance for each detection. The tracker for each bounding box is then updated per frame and is deleted when the tracker fails.
```
[the remark this answers is not in the document]
Comment (jovi-s) [reply docs/source/use_cases/multi_object_tracking.md:35]: Will implement this suggestion

File: peekingduck/pipeline/nodes/dabble/utils/tracking_files/iou_tracker/iou_tracker.py (hunk 3 above)
Comment (jovi-s) [reply peekingduck/pipeline/nodes/dabble/utils/tracking_files/iou_tracker/iou_tracker.py:36]: Changed

File: peekingduck/pipeline/nodes/dabble/utils/tracking_files/iou_tracker/track.py (hunk 4 above)
Comment (jovi-s) [reply peekingduck/pipeline/nodes/dabble/utils/tracking_files/iou_tracker/track.py:44]: Changed

File: tests/use_cases/systest_object_tracking.yml (hunk 8 above)
Comment (jovi-s) [reply tests/use_cases/systest_object_tracking.yml:11]: Will implement this suggestion

File: peekingduck/pipeline/nodes/dabble/utils/tracking_files/iou_tracker/track.py (hunk 5 above)
Comment (jovi-s) [reply peekingduck/pipeline/nodes/dabble/utils/tracking_files/iou_tracker/track.py:171]: Function removed

File: peekingduck/pipeline/nodes/dabble/utils/tracking_files/opencv_tracking.py (hunk 10 above)
Comment (jovi-s) [reply peekingduck/pipeline/nodes/dabble/utils/tracking_files/opencv_tracking.py:157]: Removed redundant tracker types

File: requirements.txt (hunk 9 above)
Comment (jovi-s) [reply requirements.txt:2]: opencv-python contains main modules for the package.
opencv-contrib-python contains both main modules and contrib/extra modules

**Review by diwei-tan — APPROVED** [review]
Reviewed by Di Wei. 17 changes to files compared to latest dev branch.

**Pull request merged**